tokio/src/runtime/task/raw.rs

```diff
@@ -279,10 +279,12 @@ unsafe fn poll<T: Future, S: Schedule>(ptr: NonNull<Header>) {
 unsafe fn schedule<S: Schedule>(ptr: NonNull<Header>) {
     use crate::runtime::task::{Notified, Task};
 
-    let scheduler = Header::get_scheduler::<S>(ptr);
-    scheduler
-        .as_ref()
-        .schedule(Notified(Task::from_raw(ptr.cast())));
+    let scheduler = Header::get_scheduler::<S>(ptr).as_ref();
+    let notify = Notified(Task::from_raw(ptr.cast()));
+    match scheduler {
+        Some(scheduler) => scheduler.schedule(notify),
+        None => crate::task::schedule::schedule(notify),
+    }
 }
```

Comment:
This looks incorrect. Consider this situation: An mpsc channel has the sender in runtime A, and the receiver in runtime B. Then, we send a message on the channel from runtime A. This causes the sender to call `wake` on the waker stored by the receiver, which corresponds to a task from runtime B, but we are on the thread for runtime A. That call eventually hits this function, and then the `None` branch of the match. This causes the task from runtime B to get scheduled on runtime A.

Reply:
There's a test failure due to something similar:
```
thread 'runtime::tests::task_combinations::test_combinations' panicked at 'there is no reactor running, must be called from the context of a Tokio 1.x runtime', tokio/src/task/schedule.rs:19:19
stack backtrace:
   0: rust_begin_unwind
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:593:5
   1: core::panicking::panic_fmt
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/panicking.rs:67:14
   2: core::panicking::panic_display
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/panicking.rs:150:5
   3: tokio::task::schedule::schedule
             at ./src/task/schedule.rs:19:19
   4: tokio::runtime::task::raw::schedule
             at ./src/runtime/task/raw.rs:286:17
   5: tokio::runtime::task::raw::RawTask::schedule
             at ./src/runtime/task/raw.rs:205:18
   6: tokio::runtime::task::harness::<impl tokio::runtime::task::raw::RawTask>::wake_by_ref
             at ./src/runtime/task/harness.rs:103:17
   7: tokio::runtime::task::waker::wake_by_ref
             at ./src/runtime/task/waker.rs:93:5
   8: core::task::wake::Waker::wake_by_ref
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/task/wake.rs:296:18
   9: core::ops::function::FnOnce::call_once
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/ops/function.rs:250:5
  10: tokio::sync::oneshot::Task::with_task::{{closure}}
             at ./src/sync/oneshot.rs:417:13
  11: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with
             at ./src/loom/std/unsafe_cell.rs:11:9
  12: tokio::sync::oneshot::Task::with_task
             at ./src/sync/oneshot.rs:415:9
  13: tokio::sync::oneshot::Inner<T>::complete
             at ./src/sync/oneshot.rs:1127:17
  14: tokio::sync::oneshot::Sender<T>::send
             at ./src/sync/oneshot.rs:608:13
  15: tokio::runtime::tests::task_combinations::test_combination
             at ./src/runtime/tests/task_combinations.rs:396:13
  16: tokio::runtime::tests::task_combinations::test_combinations
             at ./src/runtime/tests/task_combinations.rs:119:37
  17: tokio::runtime::tests::task_combinations::test_combinations::{{closure}}
             at ./src/runtime/tests/task_combinations.rs:68:24
  18: core::ops::function::FnOnce::call_once
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/ops/function.rs:250:5
  19: core::ops::function::FnOnce::call_once
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```
Here, the test calls `oneshot::Sender::send` while it is not inside the runtime context. This panics because the thread-local is not available.

Reply:
I got stuck on this this afternoon: The threads without context are hard to talk to the thread with context. This means the reference of scheduler have to put in something like channel, or in the tasks.

Reply:
I belive the message pass between sync threads and tokio runtime threads is a key capability, so the reference of scheduler in every task may have to be kept. To remove this reference, we may have to write a new `tokio::spawn` method like `tokio::spawn_without_....`, that is not a good way.